python/mlx/nn/initializers.py

```diff
@@ -0,0 +1,265 @@
+
+    Args:
+        mean (float, optional): Mean of the normal distribution. Default:
+          ``0.0``.
+        std (float, optional): Standard deviation of the normal distribution.
+          Default: ``1.0``.
+        dtype (Dtype, optional): The data type of the array. Default:
+          ``float32``.
+
+    Returns:
+        Callable[[array], array]: An initializer that returns an array with the
+        same shape as the input, filled with samples from a normal distribution.
+    """
+
+    def initializer(a: mx.array) -> mx.array:
+        return std * mx.random.normal(shape=a.shape, dtype=dtype) + mean
+
+    return initializer
+
+
+def uniform(
+    low: float = 0.0, high: float = 1.0, dtype: mx.Dtype = mx.float32
+) -> Callable[[mx.array], mx.array]:
+    r"""An initializer that returns random values from a uniform distribution.
+
+    Args:
+        low (float, optional): The lower bound of the uniform distribution.
+          Default: ``0.0``.
+        high (float, optional): The upper bound of the uniform distribution.
+          Default: ``1.0``
+        dtype (Dtype, optional): The data type of the array. Default: ``float32``.
+
+    Returns:
+        Callable[[array], array]: An initializer that returns an array
+        with the same shape as the input, filled with samples from a uniform
+        distribution
+
+    Example:
+
+        >>> init_fn = nn.init.uniform(low=0, high=1)
+        >>> out = init_fn(mx.zeros((2, 2))
+        >>> print(out)
+        array([[0.883935, 0.863726],
+               [0.617261, 0.417497]], dtype=float32)
+    """
+
+    def initializer(a: mx.array) -> mx.array:
+        return mx.random.uniform(low, high, a.shape, dtype=dtype)
+
+    return initializer
+
+
+def identity(dtype: mx.Dtype = mx.float32) -> Callable[[mx.array], mx.array]:
+    r"""An initializer that returns an identity matrix.
+
+    Args:
+        dtype (Dtype, optional): The data type of the array. Defaults:
+          ``float32``.
+
+    Returns:
+        Callable[[array], array]: An initializer that returns an identity
+        matrix with the same shape as the input.
+    """
+
+    def initializer(arr: mx.array) -> mx.array:
+        if arr.ndim != 2 or arr.shape[0] != arr.shape[1]:
+            raise ValueError(
+                f"The input array must be a square matrix but got shape {arr.shape}."
+            )
+        return mx.core.eye(n=arr.shape[0], dtype=dtype)
+
+    return initializer
+
+
+def _calculate_fan_in_fan_out(x):
+    if x.ndim < 2:
+        raise ValueError(
+            "Glorot / He initialization requires at least 2 dimensional input"
+            f" but input with {x.ndim} dimensions."
+        )
+
+    fan_in = x.shape[0]
+    fan_out = x.shape[1]
+    receptive_field = 1
+
+    if x.ndim > 2:
+        for d in x.shape[2:]:
+            receptive_field *= d
+
+        fan_in = fan_in * receptive_field
+        fan_out = fan_out * receptive_field
```

Comment:
We should double check this for the common cases (namely linear and conv). I think for our linear layers we do:

`x @ w.T` so the output dimension would be the first dim of `w`? Also would be good to check for convolutions which have out channels first and in channels last. So maybe the correct thing is:

`fan_in = x.shape[-1]` and `fan_out = x.shape[0]`?